// @fastify-resty/core/dist/decorators/hooks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OnTimeout = exports.OnResponse = exports.OnSend = exports.OnError = exports.PreSerialization = exports.PreHandler = exports.PreValidation = exports.PreParsing = exports.OnRequest = void 0;
function hookFactory(event) {
    return function (target, propertyKey) {
        Reflect.defineMetadata('fastify-resty:hook', { event }, target, propertyKey);
        if (Reflect.hasMetadata('fastify-resty:hooks', target)) {
            const hooks = Reflect.getMetadata('fastify-resty:hooks', target);
            hooks.add(propertyKey);
        }
        else {
            Reflect.defineMetadata('fastify-resty:hooks', new Set([propertyKey]), target);
        }
    };
}
exports.OnRequest = hookFactory('onRequest');
exports.PreParsing = hookFactory('preParsing');
exports.PreValidation = hookFactory('preValidation');
exports.PreHandler = hookFactory('preHandler');
exports.PreSerialization = hookFactory('preSerialization');
exports.OnError = hookFactory('onError');
exports.OnSend = hookFactory('onSend');
exports.OnResponse = hookFactory('onResponse');
exports.OnTimeout = hookFactory('onTimeout');
