// @fastify-resty/core/dist/decorators/entityController/routes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = (routeSchemas) => ({
    find: {
        method: 'GET',
        url: '/',
        schema: Object.assign({}, routeSchemas.find)
    },
    findOne: {
        method: 'GET',
        url: '/:id',
        schema: Object.assign({}, routeSchemas.findOne)
    },
    create: {
        method: 'POST',
        url: '/',
        schema: Object.assign({}, routeSchemas.create)
    },
    patch: {
        method: 'PATCH',
        url: '/',
        schema: Object.assign({}, routeSchemas.patch)
    },
    patchOne: {
        method: 'PATCH',
        url: '/:id',
        schema: Object.assign({}, routeSchemas.patchOne)
    },
    update: {
        method: 'PUT',
        url: '/',
        schema: Object.assign({}, routeSchemas.update)
    },
    updateOne: {
        method: 'PUT',
        url: '/:id',
        schema: Object.assign({}, routeSchemas.updateOne)
    },
    remove: {
        method: 'DELETE',
        url: '/',
        schema: Object.assign({}, routeSchemas.remove)
    },
    removeOne: {
        method: 'DELETE',
        url: '/:id',
        schema: Object.assign({}, routeSchemas.removeOne)
    }
});
